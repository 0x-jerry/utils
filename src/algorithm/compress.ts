import { zlibSync, unzlibSync } from 'fflate'
import { isWeb } from '../utils'

const encoder = /*@__PURE__*/ new TextEncoder()
const decoder = /*@__PURE__*/ new TextDecoder()

/**
 * Use [fflate](https://github.com/101arrowz/fflate/) to compress text
 *
 * @param text
 * @returns
 */
export function compressText(text: string): string {
  const buffer = encoder.encode(text)

  const compressed = zlibSync(buffer)

  const base64 = isWeb
    ? window.btoa(String.fromCharCode(...compressed))
    : Buffer.from(compressed).toString('base64')

  return base64
}

/**
 * Use [fflate](https://github.com/101arrowz/fflate/) to decompress text
 *
 * @param compressedText
 * @returns
 */
export function decompressText(compressedText: string): string {
  const buffer = isWeb ? base64ToUint8(compressedText) : Buffer.from(compressedText, 'base64')

  const restored = unzlibSync(buffer)

  return decoder.decode(restored)
}

function base64ToUint8(b64: string) {
  const binaryString = window.atob(b64)

  const uint8Array = new Uint8Array(binaryString.length)

  for (let i = 0; i < binaryString.length; i++) {
    uint8Array[i] = binaryString.charCodeAt(i)
  }

  return uint8Array
}
