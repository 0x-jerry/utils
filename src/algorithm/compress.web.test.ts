// @vitest-environment happy-dom
import { compressText, decompressText } from './compress'

describe('compress / decompress text', () => {
  it('should compress and decompress', async () => {
    const txt = 'text123456'

    const compressed = await compressText(txt)

    expect(compressed).toMatchSnapshot()

    const decompressed = await decompressText(compressed)

    expect(decompressed).toBe(txt)
  })
})
