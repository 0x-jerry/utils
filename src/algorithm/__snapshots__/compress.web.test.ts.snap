// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`compress / decompress text > should compress and decompress 1`] = `"eJwrSa0oMTQyNjE1AwATMwL7"`;
